pylint/utils.py

```diff
@@ -359,7 +359,7 @@ def add_message(self, msg_descr, line=None, node=None, args=None, confidence=UND
 
         If provided, the message string is expanded using args
 
-        AST checkers should must the node argument (but may optionally
+        AST checkers should provide the node argument (but may optionally
```

Comment:
I assume that you modified existing `add_message` calls due to this documentation entry. From your case it's clear that 

I'd suggest an entry:

> All emitted messages shall provide at minimum either line in which violation occurred or AST node in which violation occurred. Passing both values is also accepted.

Feel free to rephrase if you have better idea.

Reply:
Actually no, it doesn't work like this :(
If the checker inherits `IRawChecker` or `ITokenChecker`, a `line` parameter **must** be passed in `.add_message`, else there **must** be a `node` passed:
https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L153
https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L374

Reply:
Unfortunately I did not picked that up when I did review yesterday.

I'm not familiar with this part of the repository, so I cannot give a straightforward answer right now. I'd suggest extracting a new class for your check and wrap everything up so all tests pass (without modifying preexisting test suite). Then we can wait for @PCManticore input on how to tackle this hybrid AST/token checker in terms of emitting messages from them and documenting those checkers in our docs.